Lib/urllib/parse.py

```diff
@@ -535,7 +535,10 @@ def urljoin(base, url, allow_fragments=True):
     if not path and not params:
         path = bpath
         params = bparams
-        if not query:
+        # since urlparse doesn't leave any evidence of whether there was a bare
+        # '?' with an empty query string, we need to check whether it was there.
+        has_empty_query = url[0] == '?' or url.startswith(scheme + ':?')
```

Comment:
What is this `url.startswith(scheme + ':?')` special case requirement for and where is this referenced?

The behavior in Ruby and Golang was different for this scenario (but consistent).

```
require 'uri'

base_url = 'https://www.example.com/?a=b'
relative_url = 'https:?'

url = URI.join(base_url, relative_url).to_s
puts url
```

> https:?

And with golang https://go.dev/play/p/lui16M9pFyo

```
package main

import (
	"fmt"
	"net/url"
)

func main() {
	base, _ := url.Parse("https://example.com/?a=b")
	u, _ := url.Parse("http:?")
	fmt.Println(base.ResolveReference(u))
}
```

> http:?

Reply:
If this condition is removed ` or url.startswith(scheme + ':?')` from this patch, we can consider this PR as it brings the expected behavior seen across other language libraries.

Reply:
From https://datatracker.ietf.org/doc/html/rfc3986#section-5.2.2:
```
      -- A non-strict parser may ignore a scheme in the reference
      -- if it is identical to the base URI's scheme.
```
For now, `urllib.parse` behaves as a non-strict parser (there is special test for this). We can add an option to switch this, but this is a different feature.

But testing only `url.startswith(scheme + ':?')` is not enough, because `http:?#z` should clear query as well. And there are similar cases with other empty components. I am working on larger and more general PR.